File: Misc/102_lc_1334.py
'''
1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance
'''

import heapq
from typing import List
class Solution:
    def getMinimumCitiesReachable(self, city, gr, distanceThreshold, n):
        resArray = [float("inf") for _ in range(n)]
        reachableCity = [False for _ in range(n)]
        res = 0
        pq = []
        heapq.heapify(pq)
        heapq.heappush(pq, [0, city])
        resArray[city] = 0

        while pq:
            distanceTravelled, node = heapq.heappop(pq)
            for it in gr[node]:
                neighbourNode = it[0]
                weight = it[1]
                if distanceTravelled + weight < resArray[neighbourNode]:
                    resArray[neighbourNode] = distanceTravelled + weight
                    if not reachableCity[neighbourNode] and resArray[neighbourNode] <= distanceThreshold: 
                        res += 1
                        reachableCity[neighbourNode] = True
                    heapq.heappush(pq, [resArray[neighbourNode], neighbourNode])
        return res

    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        gr = {}
        for edge in edges:
            if not gr.get(edge[0]): gr[edge[0]] = [[edge[1], edge[2]]]
            else: gr[edge[0]].append([edge[1], edge[2]])

            if not gr.get(edge[1]): gr[edge[1]] = [[edge[0], edge[2]]]
            else: gr[edge[1]].append([edge[0], edge[2]])
        
        minCitiesReachable = 1_000_000_000
        res = 0
        for city in range(n):
            citiesReachable = self.getMinimumCitiesReachable(city, gr, distanceThreshold, n) if city in gr else 0
            minCitiesReachable = min(minCitiesReachable , citiesReachable)
            if citiesReachable <= minCitiesReachable:
                res = city

        return res

File: Misc/102_lc_1334.java
package Misc;

/*
 * 1334. Find the City With the Smallest Number of Neighbors at a Threshold Distance
 */

import java.util.*;

class Solution {
    private int getMinimumCitiesReachable(int city, Map<Integer, List<int[]>> gr, int distanceThreshold, int n) {
        int[] resArray = new int[n];
        Arrays.fill(resArray, Integer.MAX_VALUE);
        boolean[] reachableCity = new boolean[n];
        int res = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[] {0,city});
        resArray[city] = 0;

        while (!pq.isEmpty()) {
            int[] first = pq.poll();
            int totalDistance = first[0];
            int node = first[1];

            for (int[] it: gr.get(node)) {
                int neighbour = it[0];
                int weight = it[1];

                if (totalDistance + weight < resArray[neighbour]) {
                    resArray[neighbour] = totalDistance + weight;
                    if (!reachableCity[neighbour] && resArray[neighbour] <= distanceThreshold) {
                        res++;
                        reachableCity[neighbour] = true;
                    }
                    pq.offer(new int[] {resArray[neighbour], neighbour});
                }
            }
        }
        return res;
    }

    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        Map<Integer, List<int[]>> gr = new HashMap<>();
        for (int[] edge : edges) {
            gr.computeIfAbsent(edge[0], v -> new ArrayList<>()).add(new int[] {edge[1], edge[2]});
            gr.computeIfAbsent(edge[1], v -> new ArrayList<>()).add(new int[] {edge[0], edge[2]});
        }
        
        int minCitiesReachable = Integer.MAX_VALUE;
        int res = 0;
        for (int city=0; city<n; city++) {
            int citiesReachable = gr.containsKey(city) ? getMinimumCitiesReachable(city, gr, distanceThreshold, n) : 0;
            if (citiesReachable < minCitiesReachable) {
                minCitiesReachable = citiesReachable;
                res = city;
            } else if (citiesReachable == minCitiesReachable) {
                res = city;
            }
        }

        return res;
    }
}